features/mbedtls/platform/src/platform_alt.c

```diff
@@ -20,38 +20,33 @@
 
 #include "mbedtls/platform.h"
 #if defined(MBEDTLS_PLATFORM_SETUP_TEARDOWN_ALT)
-#include "sns_silib.h"
+mbedtls_platform_context ctx = {0};
 
-/* once https://github.com/ARMmbed/mbedtls/issues/1200 will be supported,
- * rndState should be part of mbedtls_platform_context
- * Until then, we should keep it global and extern */
-
-CRYS_RND_State_t     rndState = { { 0 } } ;
-CRYS_RND_WorkBuff_t  rndWorkBuff = { { 0 } } ;
-
-
-int mbedtls_platform_setup( mbedtls_platform_context *ctx )
+int mbedtls_platform_setup( mbedtls_platform_context *obsolete_ctx )
 {
     int ret = 0;
-    if( ctx == NULL )
-        return ( -1 );
 
-    /* call platform specific code to setup CC driver*/
-    if( ( ret = cc_platform_setup( &ctx->platform_impl_ctx ) ) != 0 )
-        return ( ret );
+    ctx.reference_count++;
 
-    if( SaSi_LibInit( &rndState, &rndWorkBuff ) != 0 )
-        return ( -1 );
-    return ( 0 );
+    if( ctx.reference_count == 1 )
+    {
+        /* call platform specific code to setup crypto driver*/
+        ret = crypto_platform_setup( &ctx.platform_impl_ctx );
+    }
+    return ( ret );
 }
 
-void mbedtls_platform_teardown( mbedtls_platform_context *ctx )
+void mbedtls_platform_teardown( mbedtls_platform_context *obsolete_ctx )
 {
-    if( ctx == NULL )
-        return;
 
-    SaSi_LibFini( &rndState );
-    cc_platform_terminate( &ctx->platform_impl_ctx );
+    ctx.reference_count--;
+
+    if( ctx.reference_count <= 0 )
+    {
+        /* call platform specific code to terminate crypto driver*/
+        crypto_platform_terminate( &ctx.platform_impl_ctx );
```

Comment:
This can call terminate multiple times if a client calls teardown too often. Do we care? Maybe this should only call terminate when reference_count goes from 1 to 0, and does nothing otherwise.

We might also want to use atomic decrement if we care about overlapping calls to `mbedtls_platform_teardown()`.

Reply:
This is what I did in the begining. See https://github.com/ARMmbed/mbed-os/pull/7099#discussion_r192667729

>Do we care?

I don't think so, because basically it's an on\off button, us pressing the off button more than once. The likelihood for that is minor, as we do set it to 0 after termination